net/rpc/server_test.go

```diff
@@ -455,6 +469,56 @@ func TestServeRequest(t *testing.T) {
 	testServeRequest(t, newServer)
 }
 
+func TestServeRequestWithInterceptor(t *testing.T) {
+	beforeHandler := 1
+	afterHandler := 3
+
+	interceptor := ServerServiceCallInterceptor(func(reqServiceMethod string, argv, replyv reflect.Value, handler func()) {
```

Comment:
We do assert the side effects after the call on two vars:

* `beforeHandler` and `afterHandler`

In the body of the interceptor, we also test that the arguments are passed correctly in correctly, both type and values.

The code in the interceptor is not actually executed up until we call `testServeRequest(t, newServer)` on line 511 which makes some actual RPC requests.

Does this help clarify the test? I can add more comments if that helps regarding the set up.

Reply:
This case:
```
1. test service call with no interceptor
```
is actually taken care on the vanilla request checking tests.

Reply:
Also, I'd like to point out that a user of this library either has a server with an interceptor or not. 

We can't specify things like "use an interceptor for this service call only" since it's not part of the current design/ approach.

Reply:
> We do assert the side effects after the call on two vars:
> 
>     * `beforeHandler` and `afterHandler`
I think the "before" and "after" distinction aren't valuable; you're setting `var x = 2` then asserting for it afterwards - this assertion will essentially never fail.

You can simplify with `var intercepted bool` because at the end of the day it either was updated or it didn't.

In addition to mutating variables local to the test, I'm interested in seeing the interceptor mutate arguments passed in, e.g)
```
server
{7,8} => {15}

server w interceptor
{7,8} => {8*,8} => {16} // is this even possible?
```

Reply:
I think the latter question is a blocker until the intended behavior (can an interceptor modify the response by mutating the request args) is documented or tested

Reply:
> is a blocker until the intended behavior (can an interceptor modify the response by mutating the request args) is documented and tested

as chatted yesterday offline, we can definitely document! but preventing mutation by design may be hard!

As you pointed out Chris, I think this is a pretty good set up regarding mutation testing:

* https://github.com/hashicorp/consul-net-rpc/compare/add-intercept-support...dnephin/can-interceptor-mutate

I can work with Daniel to build on that or add it in.